// @azure/keyvault-secrets/dist-esm/src/lro/delete/operation.js
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
import * as tslib_1 from "tslib";
/**
 * @summary Reaches to the service and updates the delete secret's poll operation.
 * @param [options] The optional parameters, which are an abortSignal from @azure/abort-controller and a function that triggers the poller's onProgress function.
 */
function update(options) {
    if (options === void 0) { options = {}; }
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        var state, name, client, requestOptions, deletedSecret, _a, error_1;
        return tslib_1.__generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    state = this.state;
                    name = state.name, client = state.client;
                    requestOptions = state.requestOptions || {};
                    if (options.abortSignal) {
                        requestOptions.abortSignal = options.abortSignal;
                    }
                    if (!!state.isStarted) return [3 /*break*/, 2];
                    return [4 /*yield*/, client.deleteSecret(name, requestOptions)];
                case 1:
                    deletedSecret = _b.sent();
                    state.isStarted = true;
                    state.result = deletedSecret;
                    if (!deletedSecret.properties.recoveryId) {
                        state.isCompleted = true;
                    }
                    _b.label = 2;
                case 2:
                    if (!!state.isCompleted) return [3 /*break*/, 6];
                    _b.label = 3;
                case 3:
                    _b.trys.push([3, 5, , 6]);
                    _a = state;
                    return [4 /*yield*/, client.getDeletedSecret(name, { requestOptions: requestOptions })];
                case 4:
                    _a.result = _b.sent();
                    state.isCompleted = true;
                    return [3 /*break*/, 6];
                case 5:
                    error_1 = _b.sent();
                    if (error_1.statusCode === 403) {
                        // At this point, the resource exists but the user doesn't have access to it.
                        state.isCompleted = true;
                    }
                    else if (error_1.statusCode !== 404) {
                        state.error = error_1;
                        state.isCompleted = true;
                    }
                    return [3 /*break*/, 6];
                case 6: return [2 /*return*/, makeDeleteSecretPollOperation(state)];
            }
        });
    });
}
/**
 * @summary Reaches to the service and cancels the secret's operation, also updating the secret's poll operation
 * @param [options] The optional parameters, which is only an abortSignal from @azure/abort-controller
 */
function cancel(_) {
    if (_ === void 0) { _ = {}; }
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        return tslib_1.__generator(this, function (_a) {
            throw new Error("Canceling the deletion of a secret is not supported.");
        });
    });
}
/**
 * @summary Serializes the create secret's poll operation
 */
function toString() {
    return JSON.stringify({
        state: this.state
    });
}
/**
 * @summary Builds a create secret's poll operation
 * @param [state] A poll operation's state, in case the new one is intended to follow up where the previous one was left.
 */
export function makeDeleteSecretPollOperation(state) {
    return {
        state: tslib_1.__assign({}, state),
        update: update,
        cancel: cancel,
        toString: toString
    };
}
