// @azure/keyvault-secrets/dist-esm/src/lro/recover/operation.js
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
import * as tslib_1 from "tslib";
/**
 * @summary Reaches to the service and updates the delete secret's poll operation.
 * @param [options] The optional parameters, which are an abortSignal from @azure/abort-controller and a function that triggers the poller's onProgress function.
 */
function update(options) {
    if (options === void 0) { options = {}; }
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        var state, name, client, requestOptions, _a, _1, _b, _c, error_1;
        return tslib_1.__generator(this, function (_d) {
            switch (_d.label) {
                case 0:
                    state = this.state;
                    name = state.name, client = state.client;
                    requestOptions = state.requestOptions || {};
                    if (options.abortSignal) {
                        requestOptions.abortSignal = options.abortSignal;
                    }
                    if (!!state.isStarted) return [3 /*break*/, 6];
                    _d.label = 1;
                case 1:
                    _d.trys.push([1, 3, , 4]);
                    _a = state;
                    return [4 /*yield*/, client.getSecret(name, { requestOptions: requestOptions })];
                case 2:
                    _a.result = (_d.sent()).properties;
                    state.isCompleted = true;
                    return [3 /*break*/, 4];
                case 3:
                    _1 = _d.sent();
                    return [3 /*break*/, 4];
                case 4:
                    if (!!state.isCompleted) return [3 /*break*/, 6];
                    _b = state;
                    return [4 /*yield*/, client.recoverDeletedSecret(name, { requestOptions: requestOptions })];
                case 5:
                    _b.result = _d.sent();
                    state.isStarted = true;
                    _d.label = 6;
                case 6:
                    if (!!state.isCompleted) return [3 /*break*/, 10];
                    _d.label = 7;
                case 7:
                    _d.trys.push([7, 9, , 10]);
                    _c = state;
                    return [4 /*yield*/, client.getSecret(name, { requestOptions: requestOptions })];
                case 8:
                    _c.result = (_d.sent()).properties;
                    state.isCompleted = true;
                    return [3 /*break*/, 10];
                case 9:
                    error_1 = _d.sent();
                    if (error_1.statusCode === 403) {
                        // At this point, the resource exists but the user doesn't have access to it.
                        state.isCompleted = true;
                    }
                    else if (error_1.statusCode !== 404) {
                        state.error = error_1;
                        state.isCompleted = true;
                    }
                    return [3 /*break*/, 10];
                case 10: return [2 /*return*/, makeRecoverDeletedSecretPollOperation(state)];
            }
        });
    });
}
/**
 * @summary Reaches to the service and cancels the secret's operation, also updating the secret's poll operation
 * @param [options] The optional parameters, which is only an abortSignal from @azure/abort-controller
 */
function cancel(_) {
    if (_ === void 0) { _ = {}; }
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        return tslib_1.__generator(this, function (_a) {
            throw new Error("Canceling the deletion of a secret is not supported.");
        });
    });
}
/**
 * @summary Serializes the create secret's poll operation
 */
function toString() {
    return JSON.stringify({
        state: this.state
    });
}
/**
 * @summary Builds a create secret's poll operation
 * @param [state] A poll operation's state, in case the new one is intended to follow up where the previous one was left.
 */
export function makeRecoverDeletedSecretPollOperation(state) {
    return {
        state: tslib_1.__assign({}, state),
        update: update,
        cancel: cancel,
        toString: toString
    };
}
